ring/consul_client.go

```diff
@@ -206,6 +215,67 @@ func (c *consulClient) WatchPrefix(prefix string, factory func() interface{}, do
 	}
 }
 
+// WatchKey will watch a given key in consul for changes. When the value
+// under said key changes, the f callback is called with the deserialised
+// value. To construct the deserialised value, a factory function should be
+// supplied which generates an empty struct for WatchKey to deserialise
+// into. Values in Consul are assumed to be JSON. This function blocks until
+// the done channel is closed.
+func (c *consulClient) WatchKey(key string, factory InstanceFactory, done <-chan struct{}, f func(interface{}) bool) {
+	const (
+		initialBackoff = 1 * time.Second
+		maxBackoff     = 1 * time.Minute
+	)
+	var (
+		backoff = initialBackoff / 2
+		index   = uint64(0)
+	)
+	for {
+		select {
+		case <-done:
+			return
+		default:
+		}
+		kvp, meta, err := c.kv.Get(key, &consul.QueryOptions{
+			RequireConsistent: true,
+			WaitIndex:         index,
+			WaitTime:          longPollDuration,
+		})
+		if err != nil {
+			log.Errorf("Error getting key %s: %v", key, err)
+			select {
+			case <-done:
+				return
+			case <-time.After(backoff):
+				backoff = backoff * 2
+				if backoff > maxBackoff {
+					backoff = maxBackoff
+				}
+				continue
+			}
+		}
+		backoff = initialBackoff
+		// Skip if the index is the same as last time, because the key value is
+		// guaranteed to be the same as last time
+		if index == meta.LastIndex {
+			continue
+		}
+		index = meta.LastIndex
+
+		var out interface{}
+		if kvp != nil {
+			out = factory()
+			if err := json.NewDecoder(bytes.NewReader(kvp.Value)).Decode(out); err != nil {
+				log.Errorf("Error deserialising %s: %v", kvp.Key, err)
+				continue
+			}
+		}
+		if !f(out) {
+			return
+		}
+	}
```

Comment:
Is there anything we can do about the significant amount of duplication between these two methods?